### PR TITLE
SSH CA enhancements

### PR DESCRIPTION
**Pull request by vishalnayak**

### CODE REVIEW

File: builtin/logical/ssh/path_config_ca.go [hunk 1]
```diff
@@ -12,6 +12,13 @@ import (
 	"golang.org/x/crypto/ssh"
 )
 
+const (
+	CAPublicKeyStoragePath            = "config/ca_public_key"
+	CAPublicKeyStoragePathDeprecated  = "public_key"
+	CAPrivateKeyStoragePath           = "config/ca_private_key"
+	CAPrivateKeyStoragePathDeprecated = "config/ca_bundle"
```
Comment (briankassouf) [inline builtin/logical/ssh/path_config_ca.go]: I don't think these need to be exported

File: builtin/logical/ssh/path_config_ca.go [hunk 2]
```diff
@@ -46,15 +53,60 @@ For security reasons, the private key cannot be retrieved later.`,
...
+	if err := req.Storage.Delete(CAPublicKeyStoragePath); err != nil {
 		return nil, err
 	}
 	return nil, nil
 }
 
+func caKey(storage logical.Storage, keyType string) (string, error) {
+	var path, deprecatedPath string
+	switch keyType {
+	case "private":
```
Comment (briankassouf) [inline builtin/logical/ssh/path_config_ca.go]: can these be made into constants?

File: builtin/logical/ssh/path_config_ca.go [hunk 3]
```diff
@@ -112,39 +164,37 @@ func (b *backend) pathConfigCAUpdate(req *logical.Request, data *framework.Field
...
 
-	privateKeyEntry, err := req.Storage.Get("config/ca_bundle")
+	storedPrivateKey, err := caKey(req.Storage, "private")
 	if err != nil {
-		return nil, fmt.Errorf("failed while reading ca_bundle: %v", err)
+		return nil, fmt.Errorf("failed to read CA private key: %v", err)
 	}
 
-	if publicKeyEntry != nil || privateKeyEntry != nil {
+	if storedPublicKey != "" || storedPrivateKey != "" {
```
Comment (briankassouf) [inline builtin/logical/ssh/path_config_ca.go]: Is it possible to get into a state where only one of these is set? If so can we protect against that happening?

File: builtin/logical/ssh/path_fetch.go [hunk 4]
```diff
@@ -19,19 +19,19 @@ func pathFetchPublicKey(b *backend) *framework.Path {
...
 
 func (b *backend) pathFetchPublicKey(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
-	entry, err := req.Storage.Get("config/ca_public_key")
+	storedPublicKey, err := caKey(req.Storage, "public")
 	if err != nil {
 		return nil, err
 	}
 
-	if entry == nil {
+	if storedPublicKey == "" {
```
Comment (briankassouf) [inline builtin/logical/ssh/path_fetch.go]: Should this return a 404?

File: builtin/logical/ssh/path_config_ca.go (hunk 1 above)
Comment (vishalnayak) [inline builtin/logical/ssh/path_config_ca.go]: done

File: builtin/logical/ssh/path_config_ca.go (hunk 3 above)
Comment (vishalnayak) [inline builtin/logical/ssh/path_config_ca.go]: This is a good point. While storing the keys, if we are able to store the first key and fail to store the next, I'm cleaning up the first one. Thanks for suggesting this.

File: builtin/logical/ssh/path_fetch.go (hunk 4 above)
Comment (vishalnayak) [inline builtin/logical/ssh/path_fetch.go]: We don't usually. Its consistent across all the API.

**Pull request merged**